### PR TITLE
Developing while writing - part 2 FindCommonBooks

### PR DESCRIPTION
**Pull request by doniacld**
## Changes
- Add `findCommonBooks` function
- Add `booksCount` function
- Add `sortBooks` function
- Test `booksCount` and `findCommonBooks`

### CODE REVIEW

File: chapter-06/2_common_books/bookworm_internal_test.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+			}
+
+			if !equalBookworms(got, testCase.want) {
+				t.Fatalf("different result: got %v, expected %v", got, testCase.want)
+			}
+		})
+	}
+}
+
+// equalBookworms is a helper to test the equity of two list of Bookworms.
```
Comment (doniacld) [inline chapter-06/2_common_books/bookworm_internal_test.go:60]:
```suggestion
// equalBookworms is a helper to test the equality of two lists of Bookworms.
```

File: chapter-06/2_common_books/bookworm.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	// Register all books on shelves.
+	booksOnShelves := booksCount(bookworms)
+
+	// List containing all the books that were read by at least 2 bookworms.
+	var booksReadBySeveralBookworms []Book
+
+	// Find books that were added to shelve more than once.
+	for book, count := range booksOnShelves {
+		if count > 1 {
+			booksReadBySeveralBookworms = append(booksReadBySeveralBookworms, book)
```
Comment (ablqk) [inline chapter-06/2_common_books/bookworm.go:52]: bookHandlerFactoryProvider

File: chapter-06/2_common_books/bookworm.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	return sortBooks(booksReadBySeveralBookworms)
+}
+
+// booksCount registers all the books and their occurrences from the bookworms shelves.
+func booksCount(bookworms []Bookworm) map[Book]uint {
+	count := make(map[Book]uint)
+
+	for _, bookworm := range bookworms {
+		for _, book := range bookworm.Books {
+			count[book]++
```
Comment (floppyzedolfin) [inline chapter-06/2_common_books/bookworm.go:66]: If a bookworm has two copies, that counts as two.

**Review by floppyzedolfin — APPROVED** [review]

File: chapter-06/1_load/bookworm.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+// loadBookworms reads the file and returns the list of bookworms, and their beloved books, found therein.
+func loadBookworms(filePath string) ([]Bookworm, error) {
+	f, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Initialise the type in which the file will be decoded.
```
Comment (ablqk) [inline chapter-06/1_load/bookworm.go:28]:
```suggestion
	// Initialise the variable in which the file will be decoded.
```

**Review by ablqk — APPROVED** [review]

File: chapter-06/1_load/bookworm.go (hunk 4 above)
Comment (floppyzedolfin) [reply chapter-06/1_load/bookworm.go:28]: Declare is fine

File: chapter-06/1_load/bookworm_internal_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		"invalid JSON": {
+			bookwormsFile: "testdata/invalid.json",
+			want:          nil,
+			wantErr:       true,
+		},
+	}
+	for name, testCase := range tests {
+		t.Run(name, func(t *testing.T) {
+			got, err := loadBookworms(testCase.bookwormsFile)
+			if err != nil && !testCase.wantErr {
```
Comment (floppyzedolfin) [inline chapter-06/1_load/bookworm_internal_test.go:42]: use a switch ?

File: chapter-06/1_load/bookworm_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	for i := range bookworms {
+		// Verify the name of the Bookworm.
+		if bookworms[i].Name != target[i].Name {
+			return false
+		}
+
+		// Verify the content of the collections of Books for each Bookworm.
+		for j := range bookworms[i].Books {
```
Comment (floppyzedolfin) [inline chapter-06/1_load/bookworm_internal_test.go:71]: chceck legnths of `.Books` !

File: chapter-06/2_common_books/bookworm_internal_test.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+			if !equalBookworms(t, got, testCase.want) {
+				t.Fatalf("different result: got %v, expected %v", got, testCase.want)
+			}
+		})
+	}
+}
+
+// equalBookworms is a helper to test the equality of two lists of Bookworms.
+func equalBookworms(t *testing.T, bookworms, target []Bookworm) bool {
+	t.Helper()
```
Comment (floppyzedolfin) [inline chapter-06/2_common_books/bookworm_internal_test.go:61]: Do we cover this ?

File: chapter-06/inprogress/bookworm_internal_test.go [hunk 8]
```diff
@@ -66,13 +66,12 @@ func TestFindCommonBooks(t *testing.T) {
...
 		want  []Book
 	}{
-		// TODO FIX ME
 		"no common book": {
 			input: []Bookworm{
 				{Name: "Fadi", Books: []Book{handmaidsTale, theBellJar}},
 				{Name: "Peggy", Books: []Book{oryxAndCrake, janeEyre}},
 			},
-			want: []Book{},
+			want: nil,
```
Comment (floppyzedolfin) [inline chapter-06/inprogress/bookworm_internal_test.go:74]: Do we explain what the differences are?

File: chapter-06/2_common_books/bookworm_internal_test.go (hunk 7 above)
Comment (doniacld) [reply chapter-06/2_common_books/bookworm_internal_test.go:61]: Very briefly and we mentioned that we will explain it later in the book.

File: chapter-06/inprogress/bookworm_internal_test.go (hunk 8 above)
Comment (doniacld) [reply chapter-06/inprogress/bookworm_internal_test.go:74]: I think we talk about the empty value...

File: chapter-06/1_load/bookworm_internal_test.go (hunk 6 above)
Comment (doniacld) [reply chapter-06/1_load/bookworm_internal_test.go]: Fixed.

File: chapter-06/1_load/bookworm_internal_test.go (hunk 5 above)
Comment (doniacld) [reply chapter-06/1_load/bookworm_internal_test.go]: Fixed.

File: chapter-06/2_common_books/bookworm.go (hunk 2 above)
Comment (doniacld) [reply chapter-06/2_common_books/bookworm.go]: I kept `commonBooks` if that's okay.

File: chapter-06/1_load/bookworm_internal_test.go [hunk 9]
```diff
@@ -39,23 +39,22 @@ func TestLoadBookworms_Success(t *testing.T) {
...
-			if err != nil && !testCase.wantErr {
-				t.Fatalf("expected an error %s, got an empty one", err.Error())
-			}
 
-			if err == nil && testCase.wantErr {
+			switch {
+			case err != nil && !testCase.wantErr:
+				t.Fatalf("expected an error %s, got an empty one", err.Error())
+			case err == nil && testCase.wantErr:
 				t.Fatalf("expected no error, got one %s", err.Error())
```
Comment (floppyzedolfin) [inline chapter-06/1_load/bookworm_internal_test.go:47]: we are missing the case where we expect an error and get one

**Review by ablqk — APPROVED** [review]

File: chapter-06/1_load/bookworm_internal_test.go (hunk 9 above)
Comment (ablqk) [reply chapter-06/1_load/bookworm_internal_test.go:47]: I found it makes the switch hard to read, so I refactored. The alternative was:
```
switch {
	case testCase.wantErr && err == nil:
		t.Fatalf("expected no error, got one %s", err.Error())
	case err != nil && !testCase.wantErr:
		t.Fatalf("expected an error %s, got an empty one", err.Error())
	case !testCase.wantErr && !equalBookworms(t, got, testCase.want):
		t.Fatalf("different result: got %v, expected %v", got, testCase.want)
}
```

Comment (ablqk) [reply chapter-06/1_load/bookworm_internal_test.go:47]: Actually the yews forest is no better:
```
if !testCase.wantErr {
	if err != nil {
		t.Fatalf("expected no error, got one %s", err.Error())
	}

	if !equalBookworms(t, got, testCase.want) {
		t.Fatalf("different result: got %v, expected %v", got, testCase.want)
	}
} else if err == nil {
	t.Fatalf("expected an error, got none")
}
```

File: chapter-06/2_common_books/bookworm_internal_test.go (hunk 7 above)
Comment (ablqk) [reply chapter-06/2_common_books/bookworm_internal_test.go]: Let's not put it in the chapter at all.

File: chapter-06/1_load/bookworm_internal_test.go (hunk 9 above)
Comment (floppyzedolfin) [reply chapter-06/1_load/bookworm_internal_test.go:47]:
```go
if testCase.wantErr  {
  if err == nil {
    t.Fatal("expected err, got nothing")
  }
  return
}

// we aren't expecting errors here, this should be the happy path
if err != nil {
  t.Fatalf("expected no error, got %v", err)
}
if !equalBookworms( got, testCase.want) {
  t.Fatalf("different result: got %v, expected %v", got, testCase.want)
}
```

**Review by ablqk — APPROVED** [review]

**Pull request merged**